packages/ampd-sdk/src/grpc/client.rs

```diff
@@ -55,103 +57,65 @@ pub trait Client {
-    let crypto = CryptoServiceClient::new(conn);
-    Ok(GrpcClient { blockchain, crypto })
-}
-
-#[derive(Debug, Clone, PartialEq)]
-pub struct BroadcastClientResponse {
-    pub tx_hash: String,
-    pub index: u64,
-}
-
-impl From<BroadcastResponse> for BroadcastClientResponse {
-    fn from(response: BroadcastResponse) -> Self {
-        BroadcastClientResponse {
-            tx_hash: response.tx_hash,
-            index: response.index,
-        }
-    }
-}
-
-#[derive(Debug, Clone, PartialEq)]
-pub struct ContractsAddresses {
-    pub voting_verifier: AccountId,
-    pub multisig_prover: AccountId,
-    pub service_registry: AccountId,
-    pub rewards: AccountId,
-}
-
-impl TryFrom<&ContractsResponse> for ContractsAddresses {
-    type Error = Report<Error>;
-
-    fn try_from(
-        response: &ContractsResponse,
-    ) -> core::result::Result<ContractsAddresses, Self::Error> {
-        let ContractsResponse {
-            voting_verifier,
-            multisig_prover,
-            service_registry,
-            rewards,
-        } = response;
-
-        Ok(ContractsAddresses {
-            voting_verifier: parse_addr(voting_verifier, "voting verifier")?,
-            multisig_prover: parse_addr(multisig_prover, "multisig prover")?,
-            service_registry: parse_addr(service_registry, "service registry")?,
-            rewards: parse_addr(rewards, "rewards contract")?,
-        })
+impl ManagedGrpcClient {
+    pub fn new(connection_handle: ConnectionHandle) -> Self {
+        Self { connection_handle }
     }
-}
-
-fn parse_addr(addr: &str, address_name: &'static str) -> Result<AccountId, Error> {
-    addr.parse::<AccountId>()
-        .change_context(AppError::InvalidAddress(address_name).into())
-        .attach_printable_lazy(|| addr.to_string())
-}
-
-pub enum KeyAlgorithm {
-    Ecdsa,
-    Ed25519,
-}
-
-pub struct Key {
-    pub id: nonempty::String,
-    pub algorithm: KeyAlgorithm,
-}
 
-impl From<Key> for KeyId {
-    fn from(key: Key) -> Self {
-        let algorithm = match key.algorithm {
-            KeyAlgorithm::Ecdsa => ampd_proto::Algorithm::Ecdsa,
-            KeyAlgorithm::Ed25519 => ampd_proto::Algorithm::Ed25519,
-        };
-
-        KeyId {
-            id: key.id.into(),
-            algorithm: algorithm as i32,
+    async fn get_service_clients(
+        &mut self,
+    ) -> Result<
+        (
+            BlockchainServiceClient<transport::Channel>,
+            CryptoServiceClient<transport::Channel>,
+        ),
+        Error,
+    > {
+        let connection_state = self.connection_handle.connection_receiver.borrow().clone();
+
+        match connection_state {
+            ConnectionState::Connected(channel) => {
+                let blockchain_service_client = BlockchainServiceClient::new(channel.clone());
+                let crypto_service_client = CryptoServiceClient::new(channel);
+                Ok((blockchain_service_client, crypto_service_client))
```

Comment:
maybe extract this so you avoid code duplication below

Reply:
agreed, separate connecting from creating the clients. I'm not even sure that part needs to have a dedicated function, usually a call just needs one or the other, so prophylactically creating both seems like a bit of a waste